### PR TITLE
Make ActiveJob locale aware

### PR DESCRIPTION
**Pull request by xijo**
When `#perform_later` is called the locale isn't stored on the
queue, which results in a locale reset when the job is performed.

An example of the problem:

```
I18n.locale = 'de'
HelloJob.perform_now # german message, correct
```

but

```
I18n.locale = 'de'
HelloJob.perform_later # english message, incorrect
```

This PR attaches the current I18n.locale to every job during the
serialization process. It is then restored during deserialization
and used to perform the job with the correct locale.

It falls back to the default locale if no serialized locale is
found in order to provide backward compatibility with previously
stored jobs. It is not necessary to clear the queue for the update.

### CODE REVIEW

File: activejob/test/cases/translation_test.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  teardown do
+    I18n.available_locales = [:en]
+  end
+
+  test 'it performs the job in the given locale' do
+    @job.locale = :de
+    @job.perform_now
+    assert_equal "Johannes says hallo", JobBuffer.last_value
+  end
+
```
Comment (kaspth) [inline activejob/test/cases/translation_test.rb]: :scissors: this line please

File: activejob/test/cases/translation_test.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  teardown do
+    I18n.available_locales = [:en]
+  end
+
+  test 'it performs the job in the given locale' do
+    @job.locale = :de
+    @job.perform_now
+    assert_equal "Johannes says hallo", JobBuffer.last_value
```
Comment (kaspth) [inline activejob/test/cases/translation_test.rb]: The expected value doesn't sound very german. Which makes me think the test is faulty (doesn't mean it is though). Am I missing something?

Comment (xijo) [inline activejob/test/cases/translation_test.rb]: @kaspth done, I changed it to sound a bit more german :laughing:

File: activejob/test/cases/translation_test.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  teardown do
+    I18n.available_locales = [:en]
+  end
+
+  test 'it performs the job in the given locale' do
+    @job.locale = :de
+    @job.perform_now
+    assert_equal "Johannes says Guten Tag", JobBuffer.last_value
```
Comment (kaspth) [inline activejob/test/cases/translation_test.rb]: Way easier to spot :+1:

**Comment by kaspth** [comment]
Overall this looks good to me :+1:, but I'm not pushing any buttons right now :smile:

I'll give this another look tomorrow.

**Comment by matthewd** [comment]
:+1: :heart:

I had an idea for how we could solve it specifically for the AM job, but I think this is much better -- i18n is a first-class concept in the framework, so it's appropriate that AJ maintains it for all jobs.

**Comment by kaspth** [comment]
Well said, @matthewd.

**Comment by cristianbica** [comment]
IMO the locale setting should be handled by the application and not the framework. It's pretty big to assume that a locale set in a controller action will be used in any jobs enqueued by that controller action. 
Image a case where a user action (a blog comment) would trigger a job that send emails to many subscribers of the blog post so all notifications email would go with the commenter's language? Although this is a pretty advanced usage it is a valid concern for me.
Anyway if this gets merged we would need a section in the AJ guide to be sure people clearly understands how AJ/I18n works.

**Comment by matthewd** [comment]
@cristianbica imagine if you sent those same emails from the controller directly, without AJ -- yes you might want to do something clever to get the right locale per mail, but the default/automatic behaviour will be that we stick to what you just set.

It seems arbitrary for us to declare the job boundary as a cut-off point... if you set a locale, it should stay set until you specify otherwise.

I very much disagree with your implication that the current "silently forget the locale" behaviour is so obvious as to not warrant documentation.

**Comment by cristianbica** [comment]
@matthewd I agree that `perform_now` / `perform_later` using different locale is an issue. I still think it's not the framework's responsibility to handle locale propagation.
Either path we decide to go (contrary to what my previous comment states) we __should__ document very well the AJ/I18n behaviour.

**Comment by tordans** [comment]
Hey guys, I wanted to jump to +1 the point @xijo is making. We migrated our DE+EN-App to rails 4.2 recently and tested our mail system locally and with software tests. Both fall back to delivering mails "now". We worked and deployed under the assumption that delivering "now" and "later" behave the same way. As far as I know that is the way they where introduced/documented. It was only until users forwarded us emails that we could believe and later reproduce the error. This feels really bad.
Delivering "now" and "later" should behave the same way. Since doing something "now" respects the current locale, doing it "later" should as well. It's still the applications job to switch the locale (I18n.with_locale) but the frameworks job to use this given locale, right?

**Comment by kaspth** [comment]
I'm still very much :+1: on this. But I agree with @cristianbica that we should document this. @xijo are you up for adding that? :smile:

It should probably be added in the serialization documentation and in the [Active Job Basics] guide: http://edgeguides.rubyonrails.org/active_job_basics.html

**Comment by kaspth** [comment]
@xijo Thanks, I've opened this again and will have a look later :smile:

File: activejob/test/cases/translation_test.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    @job = TranslatedHelloJob.new('Johannes')
+  end
+
+  teardown do
+    I18n.available_locales = [:en]
+  end
+
+  test 'it performs the job in the given locale' do
+    @job.locale = :de
+    @job.perform_now
```
Comment (robin850) [inline activejob/test/cases/translation_test.rb]: Sorry if I'm missing something but since the original issue was on `#perform_later`, there should certainly be a test for it, no ?

However with the following:

~~~ruby
test 'it performs the job in the given locale' do
  I18n.locale = :de
  TranslatedHelloJob.perform_later('Johannes')
  assert_equal "Johannes says Guten Tag", JobBuffer.last_value
end
~~~

the only failure is with Sucker Punch on my machine without your patch ; that seems strange.

File: guides/source/active_job_basics.md [hunk 5]
```diff
@@ -280,6 +280,18 @@ UserMailer.welcome(@user).deliver_later
 ```
 
 
+Internationalization
+--------------------
+
+Each job will use the same I18n.locale that was set while it the job was created.
```
Comment (robin850) [inline guides/source/active_job_basics.md]: There's an extra "it" here.

File: activejob/test/cases/translation_test.rb (hunk 4 above)
Comment (cristianbica) [inline activejob/test/cases/translation_test.rb]: @robin850 in the regular tests adapters are set to inline mode so when perform_later is called the adapters are executing the job immediately. SP probably still forks in inline mode. We did had some integration tests which starts adapters workers and I would test in there that the locale is set

File: guides/source/active_job_basics.md (hunk 5 above)
Comment (kaspth) [inline guides/source/active_job_basics.md]: I'd rephrase as `Each job uses the `I18n.locale` set when the job was created. Useful if you send emails asynchronously:`

File: guides/source/active_job_basics.md [hunk 6]
```diff
@@ -280,6 +280,18 @@ UserMailer.welcome(@user).deliver_later
...
+This behavior is especially useful if you send emails asynchronously.
+
+```ruby
+I18n.locale = :eo
```
Comment (kaspth) [inline guides/source/active_job_basics.md]: Add a newline after this.

File: guides/source/active_job_basics.md (hunk 5 above)
Comment (kaspth) [inline guides/source/active_job_basics.md]: Wups, remember to scope the line to 80 chars :wink:

File: guides/source/active_job_basics.md [hunk 7]
```diff
@@ -280,6 +280,18 @@ UserMailer.welcome(@user).deliver_later
...
+UserMailer.welcome(@user).deliver_later # @user receives his email in esparanto
```
Comment (kaspth) [inline guides/source/active_job_basics.md]: I'd rephrase as: `# Email will be localized to Esparanto.`

File: activejob/test/cases/translation_test.rb (hunk 4 above)
Comment (xijo) [inline activejob/test/cases/translation_test.rb]: @robin850 Thanks for having a look at it. From my point of view the `test 'it performs the job in the given locale' do` does a good job to test this behavior: it serializes, deserializes and executes the job, expecting that the outcome is localized.

Could you outline which test you'd like to see? I'm happy to add it then! :)

**Comment by xijo** [comment]
@kaspth @robin850 Thanks for your feedback, I added the wording changes.

File: activejob/test/cases/translation_test.rb (hunk 4 above)
Comment (robin850) [inline activejob/test/cases/translation_test.rb]: Yep I agree, the test does a good job, we should keep it. My comment was just about adding a regression test for the case you're describing in your commit message (i.e. with `#perform_later`). Maybe we can add a test (under `test/integration` as pointed by Cristian) along the lines of:

~~~ruby
test 'current I18n.locale is kept running #perform_later' do
  skip if adapter_is?(:inline)

  I18n.locale = :de
  TranslatedHelloJob.perform_later('Johannes')
  wait_for_jobs_to_finish_for(5.seconds)

  assert_equal "Johannes says Guten Tag", JobBuffer.last_value
end
~~~

What do you think ?

Comment (cristianbica) [inline activejob/test/cases/translation_test.rb]: @xijo be aware that JobBuffer won't work on the integration tests as the jobs are executed on another process. You can use the existing system in place to write content to a file(see activejob/test/support/integration/dummy_app_template.rb, activejob/test/support/integration/test_case_helpers.rb, activejob/test/integration/queuing_test.rb)

**Comment by kaspth** [comment]
@xijo I think this is almost ready to go. But have you had a chance to add a `perform_later` regression test like @robin850 suggested? :smile:

File: activejob/test/support/integration/dummy_app_template.rb [hunk 8]
```diff
@@ -14,7 +14,7 @@ class TestJob < ActiveJob::Base
 
   def perform(x)
     File.open(Rails.root.join("tmp/\#{x}"), "w+") do |f|
-      f.write x
+      f.write I18n.locale
```
Comment (xijo) [inline activejob/test/support/integration/dummy_app_template.rb]: @kaspth I added the integration test and used the fact that the content of the file wasn't asserted before. I thought that might be slightly better than adding a whole new job for this one test.

Comment (kaspth) [inline activejob/test/support/integration/dummy_app_template.rb]: As long as it doesn't break other tests, this could probably work. What do you think, @cristianbica @robin850?

Comment (cristianbica) [inline activejob/test/support/integration/dummy_app_template.rb]: It's ok. We're not asserting the content of the file.

File: guides/source/active_job_basics.md [hunk 9]
```diff
@@ -280,6 +280,19 @@ UserMailer.welcome(@user).deliver_later
 ```
 
 
+Internationalization
+--------------------
+
+Each job uses theI18n.localeset when the job was created. Useful if you send
```
Comment (kaspth) [inline guides/source/active_job_basics.md]: `I18n.locale` should be wrapped in backticks and have spaces around it :smile:

File: activejob/test/support/integration/dummy_app_template.rb (hunk 8 above)
Comment (kaspth) [inline activejob/test/support/integration/dummy_app_template.rb]: Cool :+1:

File: guides/source/active_job_basics.md (hunk 9 above)
Comment (xijo) [inline guides/source/active_job_basics.md]: Done :)

**Comment by kaspth** [comment]
@xijo There's now several failures in the Active Job integration tests. It looks like there was some problems with changing the locale to `de` in some tests.

**Comment by xijo** [comment]
@kaspth Sorry for the delay, it took me a while to figure out the problem. All should be green now.

**Comment by kaspth** [comment]
:runner: Travis :runner:

**Comment by kaspth** [comment]
Thanks @xijo, this is great :heart:

**Comment by sfroehler** [comment]
Thanks @xijo and @kaspth!!! We just ran into this issue and your fix is very helpful!!! :+1: :+1: :+1:

**Comment by sfroehler** [comment]
Are there any plans to backport this to 4.2?

**Comment by kaspth** [comment]
@rafaelfranca @matthewd do you consider this a fix or a feature? In other words should we backport this? I could definitely see it being useful for people migrating to Active Job.

**Comment by sfroehler** [comment]
In regards to ActionMailer deliver_now and deliver_later methods and for the same reasons @tordans already stated in his [comment](https://github.com/rails/rails/pull/20800#issuecomment-120845932): imho this is a fix.
Without looking into the code I would use deliver_now and deliver_later under the assumption that both methods behave the same way. Especially as there is no notice in the upgrading guides that they behave differently: http://guides.rubyonrails.org/upgrading_ruby_on_rails.html#action-mailer

**Comment by Sija** [comment]
@kaspth +1 for backporting this to 4.2

**Comment by kaspth** [comment]
Backported to 4.2 in https://github.com/rails/rails/commit/2519b87b339190ae44055036d1f1655938afd1c2

**Comment by xijo** [comment]
@kaspth Awesome, thanks! :+1:

File: activejob/lib/active_job/core.rb [hunk 10]
```diff
@@ -20,6 +20,9 @@ module Core
 
       # ID optionally provided by adapter
       attr_accessor :provider_job_id
+
+      # I18n.locale to be used during the job.
+      attr_accessor :locale
```
Comment (morgoth) [inline activejob/lib/active_job/core.rb]: Don't you think that this is too general name? I'm currently upgrading to Rails 4.2.4 and getting errors like:
```
NoMethodError:
  private method `locale' called for #<SomeJob:0x007f1053505900>
```

In my job, I have a private method `locale` defined

Comment (kaspth) [inline activejob/lib/active_job/core.rb]: Well, it'll be tough to fix until another release. To me locale sounds like the most appropriate name.
